snuba/datasets/discover.py

```diff
@@ -1,10 +1,43 @@
+from copy import deepcopy
+from typing import Any, MutableMapping
+
 from snuba.datasets.dataset import Dataset
+from snuba.datasets.entity import Entity
 from snuba.datasets.entities.factory import EntityKey, get_entity
+from snuba.datasets.entities.discover import (
+    detect_table,
+    EVENTS_COLUMNS,
+    TRANSACTIONS_COLUMNS,
+    DiscoverTransactionsEntity,
+)
+from snuba.query.parser import parse_query_from_entity
+
 
 # TODO: Clearly this is not the right model, I am just doing this for now
 # while I restructure the code, then I can move the entity selection logic back
 # out here
 class DiscoverDataset(Dataset):
     def __init__(self) -> None:
-        discover_entity = get_entity(EntityKey.DISCOVER)
-        super().__init__(default_entity=discover_entity)
+        self.discover_entity = get_entity(EntityKey.DISCOVER)
+        self.discover_transactions_entity = DiscoverTransactionsEntity()
+        super().__init__(default_entity=self.discover_entity)
+
+    # XXX: This is temporary code that will eventually need to be ported to Sentry
+    # since SnQL will require an entity to always be specified by the user.
+    def get_entity(self, query_body: MutableMapping[str, Any]) -> Entity:
+        # First parse the query with the default entity
+        query_copy = deepcopy(query_body)
+        query = parse_query_from_entity(query_copy, self.discover_entity)
```

Comment:
This has a few issues:
- we are parsing the query twice at least and actually every time we fetch an entity object.
- the query body is not the right source to pick the appropriate entity. That should be the  parsed query object, otherwise this will only work very early on during parsing, since the dictionary will not be available anymore after the initial parsing.
- It would not work for joins where multiple queries are involved in a query.

I think we are addressing this problem from the wrong side. There is a difference between the entities to be used for a query and the static declaration of an entity in the dataset. This method was meant to return a given entity by name from the dataset.
The entities involved in a query should be referenced (not directly otherwise you have a circular dependency) in the query object itself as we do for storages today.
So, within parsing we recognize the involved entity, and whoever has access to the query from that point on know which entities are involved.

Is this meant to be a temporary solution related to the transactions/errors split or is this meant to me permanent ?

Reply:
Since the entity selection code will move into Sentry eventually, I think the query body is the right place to pick the entity, and it should not have access to the internal Snuba query representation in order to determine the entity. But that involves parsing twice since the Snuba version of the entity selection code depends on the parsed query. I agree this isn't great even if temporary so I switched the order back to avoid this, though I think the previous sequence is the more correct one.

In relation to joins, what aspects are you concerned about specifically? It is true that joins don't work, but neither did the original version.